### PR TITLE
New scroller !

### PR DESCRIPTION
**Pull request by lorem--ipsum**
Much scrolling, very viewport.

### CODE REVIEW

File: src/client/components/scroller/scroller.tsx [hunk 1]
```diff
@@ -2,44 +2,380 @@ require('./scroller.css');
...
     this.state = {
+      scrollTop: 0,
       scrollLeft: 0,
-      scrollTop: 0
+      viewportHeight: 0,
+      viewportWidth: 0
+    };
+  }
+
+  getTopLeftCornerStyle(): React.CSSProperties {
```
Comment (vogievetsky) [inline src/client/components/scroller/scroller.tsx]: I think it would be shorter and simpler to write these functions as:

```typescript
type XPos = 'left' | 'right';
type yPos = 'top' | 'bottom';

getCornerStyle(xPos: XPos, yPos: YPos): React.CSSProperties {
  const { layout } = this.props;
  var style = {};
  if (xPos === 'left') {
    style.left = 0;
    style.width = layout.left;
  } else {
    /* ... */
  }

  if (yPos === 'top') {
    /* ... */
  } else {
    /* ... */
  }
  return style;
}
```

Then you would not need kebab-case transformer function

File: src/client/components/scroller/scroller.tsx [hunk 2]
```diff
@@ -2,44 +2,380 @@ require('./scroller.css');
...
+    const { x, y } = this.getRelativeMouseCoordinates(event);
+
+    this.props.onMouseMove(x, y);
+  }
+
+  firstUp(str: string): string {
+    return str ? str.charAt(0).toUpperCase() + str.slice(1) : undefined;
+  }
+
+  toKebabCase(str: string): string {
```
Comment (vogievetsky) [inline src/client/components/scroller/scroller.tsx]: maybe this should live in a common unfitly file like `/dev/null`

File: src/client/components/scroller/scroller.tsx [hunk 3]
```diff
@@ -2,44 +2,380 @@ require('./scroller.css');
...
+
+      <div
+        className="event-container"
+        ref="eventContainer"
+        onScroll={this.onScroll.bind(this)}
+        onClick={this.onClick.bind(this)}
+        onMouseMove={this.onMouseMove.bind(this)}
+        onMouseLeave={onMouseLeave || null}
+       >
+        <div className="event-target" style={this.getTargetStyle()}/>
```
Comment (vogievetsky) [inline src/client/components/scroller/scroller.tsx]: why is this needed?

File: src/client/components/scroller/scroller.tsx [hunk 4]
```diff
@@ -2,44 +2,380 @@ require('./scroller.css');
...
+
+    return {
+      width: layout.left,
+      left: 0,
+      top: layout.top - scrollTop,
+      bottom: layout.bottom
+    };
+  }
+
+  getTopShadowStyle(): React.CSSProperties {
```
Comment (vogievetsky) [inline src/client/components/scroller/scroller.tsx]: similarly I would write this as `getShadowStyle(side: Side): React.CSSProperties` and have a 4 way switch statement inside.

File: src/client/components/scroller/scroller.tsx (hunk 3 above)
Comment (vogievetsky) [inline src/client/components/scroller/scroller.tsx]: nvm

File: src/client/components/scroller/scroller.tsx [hunk 5]
```diff
@@ -2,44 +2,380 @@ require('./scroller.css');
...
+  getBodyOverlayStyle(): React.CSSProperties {
+    return this.getBodyStyle();
+  }
+
+  getTargetStyle(): React.CSSProperties {
+    const { layout } = this.props;
+
+    return {
+      top: 0,
+      left: 0,
```
Comment (vogievetsky) [inline src/client/components/scroller/scroller.tsx]: you do not need to set top, left here

File: src/client/components/scroller/scroller.tsx (hunk 3 above)
Comment (vogievetsky) [inline src/client/components/scroller/scroller.tsx]: Can we name is something better? Scrolling expander? Jaws of life?

http://s1.cdn.autoevolution.com/images/news/app-to-guide-the-jaws-of-life-33667_1.jpg

File: src/client/components/scroller/scroller.tsx (hunk 1 above)
Comment (lorem--ipsum) [inline src/client/components/scroller/scroller.tsx]: Good idead

File: src/client/components/scroller/scroller.tsx (hunk 2 above)
Comment (lorem--ipsum) [inline src/client/components/scroller/scroller.tsx]: `/dev/null` ?

File: src/client/components/scroller/scroller.tsx (hunk 3 above)
Comment (lorem--ipsum) [inline src/client/components/scroller/scroller.tsx]: I... `event-target` seems pretty explicit to me, don't you think so ?

File: src/client/components/scroller/scroller.tsx [hunk 6]
```diff
@@ -2,44 +2,315 @@ require('./scroller.css');
...
+    if (!(this.props.layout as any)[side]) return null; // no gutter ? no shadow.
+    if (!this.shouldHaveShadow(side)) return null;
+
+    return <div className={`${side}-shadow`} style={this.getShadowStyle(side)}/>;
+  }
+
+  renderCorner(yPos: YSide, xPos: XSide): JSX.Element {
+    var style = this.getCornerStyle(yPos, xPos);
+    var element = (this.props as any)[yPos + this.firstUp(xPos) + 'Corner'];
+
```
Comment (vogievetsky) [inline src/client/components/scroller/scroller.tsx]: adding a `if (!element) return null;` would make the rendered DOM cleaner

File: src/client/components/scroller/scroller.tsx [hunk 7]
```diff
@@ -2,44 +2,315 @@ require('./scroller.css');
...
+    this.props.onMouseMove(x, y);
+  }
+
+  firstUp(str: string): string {
+    return str ? str.charAt(0).toUpperCase() + str.slice(1) : undefined;
+  }
+
+  renderGutter(side: XSide | YSide): JSX.Element {
+    var element = (this.props as any)[`${side}Gutter`];
+
```
Comment (vogievetsky) [inline src/client/components/scroller/scroller.tsx]: same here

**Comment by vogievetsky** [comment]
gr 🎱
